### PR TITLE
[SPARK-5549] Define TaskContext interface in Scala.

### PR DESCRIPTION
**Pull request by rxin**
So the interface documentation shows up in ScalaDoc.

### CODE REVIEW

File: core/src/test/java/test/org/apache/spark/JavaTaskCompletionListenerImpl.java [hunk 1]
```diff
@@ -15,9 +15,10 @@
  * limitations under the License.
  */
 
-package org.apache.spark.util;
+package test.org.apache.spark;
```
Comment (rxin) [inline core/src/test/java/test/org/apache/spark/JavaTaskCompletionListenerImpl.java]: note that I changed the package to make sure we test package visibility correctly also. 

We should do this for other Java API tests, but we can do those later.

File: core/src/main/scala/org/apache/spark/TaskContext.scala [hunk 2]
```diff
@@ -15,112 +15,116 @@
...
-   * Whether the task has been killed.
+   * Returns true if the task has been killed.
    */
-  public abstract boolean isInterrupted();
+  def isInterrupted(): Boolean
 
-  /** @deprecated use {@link #isRunningLocally()} */
-  @Deprecated
-  public abstract boolean runningLocally();
+  /** @deprecated use { @link #isRunningLocally()}*/
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContext.scala]: Scalastyle didn't like the whitespace here.

File: core/src/main/scala/org/apache/spark/TaskContext.scala [hunk 3]
```diff
@@ -15,112 +15,116 @@
...
+  @deprecated("1.2.0", "use isRunningLocally")
+  def runningLocally(): Boolean
 
-  public abstract boolean isRunningLocally();
+  /**
+   * Returns true if the task is running locally in the driver program.
+   * @return
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/TaskContext.scala]: Can probably drop this `@return` tag.

**Comment by JoshRosen** [comment]
LGTM; this is a pretty straightforward change.  Compared to implementing this in Java, this gains the benefit of having this class's docs appear in Scaladoc.  I guess there's a potential downside of Java compatibility / usability being broken should we add new methods, but I think we can address that via our review processes.  It might be nice to add a comment at the top of the `TaskContext` and `TaskContextImpl` files referencing the `JavaTaskCompletionListenerImpl` test case so that people remember to update it when adding new methods.

**Comment by JoshRosen** [comment]
I don't know whether MiMa actually checks our Java classes, so that might be another side-benefit of porting this back to Scala.

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Going to merge since tests passed previously, and the latest failure was due to a flaky test in streaming.

**Pull request closed without merging**